bee-network/bee-gossip/src/handshake.rs

```diff
@@ -0,0 +1,380 @@
+}
+
+async fn send_handshake_request(
+    writer: &mut BufWriter<OwnedWriteHalf>,
+    to: IpAddr,
+    local_id: &LocalId,
+) -> Result<BytesMut, HandshakeError> {
+    let ty = PacketType::Handshake;
+
+    let data = HandshakeRequest::new(to).protobuf()?;
+    let signature = local_id.sign(&data).to_bytes();
+
+    let packet = Packet::new(ty, &data, local_id.public_key().as_ref(), &signature);
+    let packet_bytes = packet.protobuf().map_err(HandshakeError::Encode)?;
+
+    writer
+        .write_all(packet_bytes.as_ref())
+        .await
+        .map_err(HandshakeError::Io)?;
+    writer.flush().await.map_err(HandshakeError::Io)?;
+
+    Ok(data)
+}
+
+async fn send_handshake_response(
+    writer: &mut BufWriter<OwnedWriteHalf>,
+    req_data: &[u8],
+    local_id: &LocalId,
+) -> Result<(), HandshakeError> {
+    let ty = PacketType::Handshake;
+
+    let data = HandshakeResponse::new(req_data).protobuf()?;
+    let signature = local_id.sign(&data).to_bytes();
+
+    let packet = Packet::new(ty, &data, local_id.public_key().as_ref(), &signature);
+    let packet_bytes = packet.protobuf().map_err(HandshakeError::Encode)?;
+
+    writer
+        .write_all(packet_bytes.as_ref())
+        .await
+        .map_err(HandshakeError::Io)?;
+    writer.flush().await.map_err(HandshakeError::Io)?;
+
+    Ok(())
+}
+
+async fn await_request(
+    reader: &mut BufReader<OwnedReadHalf>,
+    writer: &mut BufWriter<OwnedWriteHalf>,
+    local_id: &LocalId,
+    _peer_config: &ManualPeerConfig,
+) -> Result<PeerId, HandshakeError> {
+    let mut buf = vec![0; MAX_HANDSHAKE_PACKET_SIZE];
+
+    let packet = loop {
+        if let Ok(num_received) = reader.read(&mut buf).await {
+            if num_received == 0 {
+                return Err(HandshakeError::ConnectionResetByPeer);
+            }
+
+            if num_received > MAX_HANDSHAKE_PACKET_SIZE {
+                return Err(HandshakeError::PacketSizeMismatch {
+                    received: num_received,
+                    max_allowed: MAX_HANDSHAKE_PACKET_SIZE,
+                });
+            }
+
+            let packet = Packet::from_protobuf(&buf[..num_received]).map_err(HandshakeError::Decode)?;
+            let packet_type = packet.ty().map_err(HandshakeError::PacketType)?;
+
+            if matches!(packet_type, PacketType::Handshake) {
+                log::info!("received handshake request.");
+
+                let req = HandshakeRequest::from_protobuf(packet.data())?;
+                let peer_req_data = req.protobuf()?;
+
+                HandshakeValidator::validate_request(&peer_req_data)?;
+
+                send_handshake_response(writer, &peer_req_data, local_id).await?;
+
+                break packet;
+            }
+        }
+    };
+
+    let peer_public_key = packet.public_key();
+    let mut pk = [0u8; 32];
+    pk.copy_from_slice(&peer_public_key[..32]);
+    let peer_public_key = ed25519::PublicKey::try_from_bytes(pk).map_err(HandshakeError::PublicKey)?;
```

Comment:
is this copy necessary?

Reply:
I realized it is necessary because all those methods receive `[u8; 32]` arrays and not references to them. 

However, I thought that maybe we are double checking bounds when calling `copy_from_slice` as the signature of that function uses a slice and not an array referece, meaning that it doesn't have the size of the slice during compilation.

So I decided to write an unsafe version of this function that calls `clone` instead and check the generated assembly in my good ol' friend [godbolt](https://rust.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:2,endLineNumber:9,positionColumn:2,positionLineNumber:9,selectionStartColumn:2,selectionStartLineNumber:9,startColumn:2,startLineNumber:9),source:'pub+fn+foo(bytes:+%26%5Bu8%5D)+-%3E+%5Bu8%3B+32%5D+%7B%0A++++let+mut+arr+%3D+%5B0%3B+32%5D%3B%0A++++arr.copy_from_slice(%26bytes%5B..32%5D)%3B%0A++++arr%0A%7D%0A%0Apub+fn+foo2(bytes:+%26%5Bu8%5D)+-%3E+%5Bu8%3B+32%5D+%7B%0A++++unsafe%7B%26*(%26bytes%5B..32%5D+as+*const+%5Bu8%5D+as+*const%5Bu8%3B+32%5D)%7D.clone()%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1520,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-O',selection:(endColumn:12,endLineNumber:40,positionColumn:1,positionLineNumber:23,selectionStartColumn:12,selectionStartLineNumber:40,startColumn:1,startLineNumber:23),source:1,tree:'1'),l:'5',n:'0',o:'rustc+1.52.0+(Rust,+Editor+%231,+Compiler+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',tree:'1',wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+1.52.0+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

The diff between the two functions is:
```diff
1c1
< example::foo:
---
> example::foo2:
3,5d2
<         xorps   xmm0, xmm0
<         movups  xmmword ptr [rdi + 16], xmm0
<         movups  xmmword ptr [rdi], xmm0
7c4
<         jbe     .LBB0_1
---
>         jbe     .LBB1_1
15,16c12,13
< .LBB0_1:
<         lea     rax, [rip + .L__unnamed_1]
---
> .LBB1_1:
>         lea     rax, [rip + .L__unnamed_2]
```
which means we are saving one `xor` and two `mov` with the unsafe version.

Reply:
I watched a few of Matt's talks, because he's also super entertaining :smile: I wonder how I haven't seen anything from him yet. I also wonder though whether we're falling into the pre-optimization trap here. :thinking:  IMHO benchmarks should first show, whether introducing `unsafe` here is really worth it. As Matt said in the talk you linked to, removing some ops in the assempbly doesn't necessarily mean, that this results in a noticable performance improvement. Especially in this situation here: Handshaking is only done once per gossip connection, maybe roughly done at a rate 1(handshake):10000(gossip messages).

Reply:
I agree that this code doesn't run a lot but this same idea of copying into an array is used in several places in the network. Maybe we should abstract this to a new safe function of type `fn<const N: usize, T>(&[T]) -> Option<&[T; N]>` in some utils crate or something.